frontend/apps/reader/readerui.lua

```diff
@@ -458,6 +458,7 @@ function ReaderUI:init()
     self:handleEvent(Event:new("DocSettingsLoad", self.doc_settings, self.document))
     -- we only read settings after all the widgets are initialized
     self:handleEvent(Event:new("ReadSettings", self.doc_settings))
+    self:handleEvent(Event:new("PageUpdate", self.doc_settings:readSetting("last_page") or 1))
```

Comment:
I don't think this (ReaderUI:init()) is really the place to do that. Doing it here will have it done also with ReaderRolling (cre/epub documents), which has no issue and so does not need it (and it might cause issues if this provokes too much stuff in readerrolling too early before anything else is set up.

It should probably be done in ReaderPaging only, and ~~probably way later than just after onReadSettings~~ maybe not, it seems ReaderPaging already does that very early (probably because it's less complex that crengine in the amount of events and their proper ordering):
https://github.com/koreader/koreader/blob/f836f6a237b7f9ac52720cc50073c2bdfc711305/frontend/apps/reader/modules/readerpaging.lua#L189-L191
and this _gotoPage seems to already send that event (?):
https://github.com/koreader/koreader/blob/f836f6a237b7f9ac52720cc50073c2bdfc711305/frontend/apps/reader/modules/readerpaging.lua#L1103-L1120

Reply:
Didn't work if done it inside of onReadSettings after the last line. I even try to call _gotoPage two times.
I even try the ReaderPaging:onGotoPage variant.
Maybe some kind of event dependency queued to be executed after onReadSettings?

Reply:
Probably. Other modules will also process onReadSettings, and may be it depends on some stuff done by omdules who does it after.
Would be nice to figure out why it does not work and understand what's needed.
There is another event/handler ReaderPaging:onReaderReady() that is called way later, that oculd be a last resort solution :)

Reply:
> Probably. Other modules will also process onReadSettings, and may be it depends on some stuff done by omdules who does it after. Would be nice to figure out why it does not work and understand what's needed. 

I'll try it when find more time, another fifty percent there we go...

Reply:
Found the reason.

The settings that are needed are here: https://github.com/koreader/koreader/blob/master/frontend/apps/reader/modules/readerzooming.lua#L267

The events order:

02/02/24-23:32:42 WARN ReaderView onPageUpdate end 
02/02/24-23:32:42 WARN ReaderPaging onPageUpdate end 
02/02/24-23:32:42 WARN ReaderZooming onPageUpdate end 
02/02/24-23:32:42 WARN ReaderPaging onReadSettings end 
02/02/24-23:32:42 WARN ReaderZooming onReadSettings 
02/02/24-23:32:42 WARN ReaderView onReaderReady 
02/02/24-23:32:42 WARN ReaderPaging onReaderReady end

ReaderZooming onReadSettings is called after ReaderPaging onReadSettings (That calls onPageUpdate).

So we use onReaderReady as the last resource, or is there a better way to control the events order? Should we put zoom onReadSettings first if it doesn't break anything?

Reply:
> Should we put zoom onReadSettings first if it doesn't break anything?

That's always a bit risky (you don't know before you try that if Zooming won't need something from Paging :) so best to not take chances and don't break what's not broken).

But,if you need stuff from Zooming, and all that you want is really related to doing the right thing for _some_ zoom/scroll modes, just have the complementary adjustments logically done in ReaderZooming:onReadSettings(), no ? Either by duplicating the little part of the work you need from onPageUpdate, or by calling onPageUpdate a second time when you need it (if there are conditions under which it's needed or not needed, it's fine if you use them here as some kind of documentation, ie
```
if not top_to_bottom then
  -- position us on the page as expected
end
```

Reply:
> > Should we put zoom onReadSettings first if it doesn't break anything?
> 
> That's always a bit risky (you don't know before you try that if Zooming won't need something from Paging :) so best to not take chances and don't break what's not broken).
> 
> But,if you need stuff from Zooming, and all that you want is really related to doing the right thing for _some_ zoom/scroll modes, just have the complementary adjustments logically done in ReaderZooming:onReadSettings(), no ? Either by duplicating the little part of the work you need from onPageUpdate, or by calling onPageUpdate a second time when you need it (if there are conditions under which it's needed or not needed, it's fine if you use them here as some kind of documentation, ie
> 
> ```
> if not top_to_bottom then
>   -- position us on the page as expected
> end
> ```

I sure do hope that we don't need to use recalculate:

--[[
This method is supposed to be only used by ReaderPaging
--]]
function ReaderView:recalculate()

https://github.com/koreader/koreader/blob/master/frontend/apps/reader/modules/readerview.lua#L675


that is called from ReaderView:onPageUpdate:
https://github.com/koreader/koreader/blob/master/frontend/apps/reader/modules/readerview.lua#L965


It's supposed to be only used by ReaderPaging, and we would be using it for ReaderZooming.


More logs:

02/03/24-07:21:17 WARN recalculate() Entered in recalculate function 
02/03/24-07:21:17 WARN recalculate() Entered in recalculate function 
02/03/24-07:21:17 WARN recalculate() Entered in recalculate function 
02/03/24-07:21:17 WARN recalculate() Entered in recalculate if self.ui.paging and self.state.page then 
02/03/24-07:21:17 WARN recalculate() About to check if self.ui.zooming.zoom_bottom_to_top then 
02/03/24-07:21:17 WARN ReaderView onPageUpdate end 
02/03/24-07:21:17 WARN ReaderPaging onPageUpdate end 
02/03/24-07:21:17 WARN recalculate() Entered in recalculate function 
02/03/24-07:21:17 WARN recalculate() Entered in recalculate if self.ui.paging and self.state.page then 
02/03/24-07:21:17 WARN recalculate() About to check if self.ui.zooming.zoom_bottom_to_top then 
02/03/24-07:21:17 WARN ReaderZooming onPageUpdate end 
02/03/24-07:21:17 WARN ReaderPaging onReadSettings end 
02/03/24-07:21:17 WARN ReaderZooming onReadSettings 
02/03/24-07:21:17 WARN ReaderView onReaderReady 
02/03/24-07:21:17 WARN ReaderPaging onReaderReady end 

Is checking things from zoom:
02/03/24-07:21:17 WARN recalculate() About to check if self.ui.zooming.zoom_bottom_to_top then 

But zoom settings are read later:
02/03/24-07:21:17 WARN ReaderZooming onReadSettings

Reply:
How about put self:setZoom() at the end of this function?

https://github.com/koreader/koreader/blob/master/frontend/apps/reader/modules/readerzooming.lua#L218

Reply:
I don't know, I can't say :) I'm not confortable with all this, and don't want to be and follow you on the adventure :)
Just pointing out how twisted and fragile all that is. And my only criteria for judging it is if _reads_ nice and logical :)
If it's too complex and cloudy, just go with onReaderReady and do the additional stuff `if not top_to_bottom or not left_to_right` (whatever the conditions for needing that additional positioning to go out of the default anchor of top left of page at top left of screen).